Switch nightly ci to use the golang version of vcr_cassette_update

iyabchen opened the pull request:
<!-- Put a description of what this PR is for here, along with any references to issues that this resolves or contributes to -->

https://github.com/hashicorp/terraform-provider-google/issues/18249

Part 1: https://github.com/GoogleCloudPlatform/magic-modules/pull/11147

This is tested in https://github.com/GoogleCloudPlatform/magic-modules/pull/11282

<!--
Please self-review your PR against the review checklist before creating it: https://googlecloudplatform.github.io/magic-modules/contribute/review-pr/

Completing the checklist will help speed up the review process, and we appreciate you spending time on them before sending
your code to be reviewed.

If your PR is still work in progress, please create it in draft mode
-->

<!-- AUTOCHANGELOG for Downstream PRs.

Please select one of the following "release-note:" headings:
    - release-note:enhancement
    - release-note:bug
    - release-note:note
    - release-note:new-resource
    - release-note:new-datasource
    - release-note:deprecation
    - release-note:breaking-change
    - release-note:none

Unless you choose release-note:none, please add a release note.

See https://googlecloudplatform.github.io/magic-modules/contribute/release-notes/ for writing good release notes.

You can add more release note blocks if you want more than one CHANGELOG
entry for this PR.
-->
**Release Note Template for Downstream PRs (will be copied)**

```release-note:none

```

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR hasn't generated any diffs, but I'll let you know if a future commit does.

shuyama1 commented on .ci/magician/cmd/vcr_cassette_update.go:
```diff
@@ -0,0 +1,210 @@
+package cmd
+
+import (
```
@iyabchen qq: I assume this is the rewrite of https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/scripts/go-plus/vcr-cassette-update/vcr_cassette_update.sh. Where do we upload the cassettes to the GCS bucket after running in recording? Specifically [this step](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/scripts/go-plus/vcr-cassette-update/vcr_cassette_update.sh#L130) in the original script.

iyabchen replied:
Yes, this is for [vcr_cassette_update.sh](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/scripts/go-plus/vcr-cassette-update/vcr_cassette_update.sh).

Good catch, I completely miss that, just updated the code and added a test for recording.

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR hasn't generated any diffs, but I'll let you know if a future commit does.

iyabchen commented:
Note that this PR depends on https://github.com/GoogleCloudPlatform/magic-modules/pull/11147, which reduces circular dependency and help to use mock for testing

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR hasn't generated any diffs, but I'll let you know if a future commit does.

shuyama1 commented on .ci/gcb-vcr-nightly.yml:
```diff
@@ -41,3 +46,5 @@ availableSecrets:
       env: SA_KEY
     - versionName: projects/673497134629/secrets/ci-test-public-advertised-prefix-description/versions/latest
       env: GOOGLE_PUBLIC_AVERTISED_PREFIX_DESCRIPTION
+    - versionName: projects/673497134629/secrets/github-classic--repo-workflow/versions/latest
```
we'll also need to add this to `secretEnv`.

shuyama1 commented:
@iyabchen I've added the necessary permissions to the SA and the job successfully pulled the cassettes and run VCR in replaying and recording mode but I seems the job failed at uploading cassettes step. The `vcr-nightly-temp-testing` in https://github.com/GoogleCloudPlatform/magic-modules/pull/11171 should point to the latest run. Please let me know if you can't view the build log/ have any question/ need help debugging the issue. Thanks!

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR hasn't generated any diffs, but I'll let you know if a future commit does.

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR hasn't generated any diffs, but I'll let you know if a future commit does.

iyabchen commented:
> @iyabchen I've added the necessary permissions to the SA and the job successfully pulled the cassettes and run VCR in replaying and recording mode but I seems the job failed at uploading cassettes step. The `vcr-nightly-temp-testing` in #11171 should point to the latest run. Please let me know if you can't view the build log/ have any question/ need help debugging the issue. Thanks!

Hi @shuyama1 ,
I tried to run it locally, and since missing some env variable, so it doesn't proceed. I suspect the recording is not working properly for similar reasons, and I adjusted the upload log steps to upload building/testing log first, then upload the cassettes later, hoping the testing log can show some errors. 
Could you copy again and let me know the results?

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR hasn't generated any diffs, but I'll let you know if a future commit does.

melinath commented on .ci/magician/cmd/tester.go:
```diff
@@ -1,9 +1,8 @@
-package vcr
+package cmd
```
Why does this need to be in the cmd package? It would be preferable to keep it separated so that it can be easily used as a library in other contexts if possible.

melinath replied:
cc @trodge

iyabchen replied:
The move is in this PR: https://github.com/GoogleCloudPlatform/magic-modules/pull/11147/files as part 1. 
It is to avoid circular dependency. 
I would like to have vcr Tester struct reference ExecRunner interface rather than the actual exec runner struct for tests.  
Since ExecRunner interface is defined in cmd package, and cmd package needs to reference tester struct, not moving will cause circular dependency. 
One alternative is to move the interface definition out of cmd.

melinath replied:
I think that moving the interface definition out would be preferable (though I'm not sure how best to make that match best practices.)

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR hasn't generated any diffs, but I'll let you know if a future commit does.

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR hasn't generated any diffs, but I'll let you know if a future commit does.

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR hasn't generated any diffs, but I'll let you know if a future commit does.

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR hasn't generated any diffs, but I'll let you know if a future commit does.

shuyama1 reviewed (approved):
LGTM! I'll remove the temp testing trigger and monitor the VCR run tonight. Thanks!

shuyama1 commented:
@iyabchen Looks like the VCR nightly run failed last night. See [logs](https://pantheon.corp.google.com/cloud-build/builds/e9676603-1717-4e3b-92df-fce83709a31d?project=graphite-docker-images&e=-13802955&mods=logs_tg_staging&invt=AbZZpQ&inv=1). Let's revert it for now?
